Add automated "caniuse" support

RReverser opened the pull request:
I see something like this was already @sbc100's TODO, but I realised that I already added a few features recently where I had to look up caniuse data by hand, and got to that state where it's tempting to automate everything 😅

This PR allows to use up-to-date compat info for JS and Wasm features instead of having to manually sprinkle `MIN_XXX_VERSION` comparisons everywhere (which in some places have been already outdated, e.g. for things that Safari implemented since).

Various caniuse or browser compat data packages vary in range ~3-12MB depending on amount of data you need, but either way that seems like too much for just the minimal browser versions we're interested in. Instead, I'm adding a script that downloads and preprocesses MDN browser compat data + webassembly.org Wasm feature data and produces a ~100KB JSON file that can be added to the repo. Then, I added JS and Python helpers to just get `caniuse` for given feature names.

This approach seems better as the data can be updated at-will and each user doesn't have to download the large raw datasets and preprocess them locally (which could also lead to all sorts of issues in case of remote changing the format or just server being down).

FWIW JSON could be made even smaller at the cost of readability, but I think it's already at reasonable size and I found it useful to be able to lookup feature names in an editor.

RReverser commented:
I'll look into the remaining failures a bit later - at least some tests seem to need explicit MIN_XXX_VERSION set to actually include polyfills as they expect - but wanted to open to see if general direction makes sense.

sbc100 commented on tools/browser_compat_data.json:
```diff
@@ -0,0 +1,5939 @@
+{
```
Can we limit this to just the features we care about.   It should be easy enough to re-generate each time we decide we care about a new feature.

It would also make it easier to eyeball / review.

RReverser replied:
We could but idk seems like an unnecessary overhead to add new features with little benefit (size is already reasonable)? Besides, this way external JS libraries can use `#if caniuse` for features they care about too, even if core doesn't.

RReverser replied:
I mean, the whole point of this feature is to avoid maintenance overhead so you can just do `#if caniuse` for any feature without having to track them somewhere one by one.

If we still need to add their names somewhere each time we want to compile-test some feature, might as well keep copy-pasting browser versions as well manually like before.

sbc100 replied:
It seems pretty easy to add your feature to the list and re-run the script json-generating script.

sbc100 replied:
I don't think the size is a big issue.    We can just checkin the whole thing.

But its a shame its hard to confirm in the review that the new and old checks are equivalent.    Can you confirm this and/or explain if there are any differences?

RReverser replied:
> But its a shame its hard to confirm in the review that the new and old checks are equivalent.

Do you mean before/after this PR? Yeah, as mentioned in description, there were already some differences in e.g. manual checks assuming that feature is not supported in Safari but it actually is supported nowadays.

That's why I think trusting the database is better - it's pretty much guaranteed to have more up-to-date information.

Also note that I intentionally limited browser compat data to JavaScript builtins (as in, ECMAScript spec things, excluding all DOM APIs) so it won't update as often as you'd think.

It's relatively rare for new JS core APIs to appear, and I'd expect to see changes only when new browser versions are coming out and implement a couple of features at a time. So if one wants, I think it would be still reasonably easy to review.

sbc100 replied:
sgtm

sbc100 commented on src/library_atomic.js:
```diff
@@ -11,10 +11,7 @@ addToLibrary({
-//   https://bugzilla.mozilla.org/show_bug.cgi?id=1467846
-// And at the time of writing, no other browser has it either.
-#if MIN_EDGE_VERSION < 91 || MIN_CHROME_VERSION < 91 || MIN_SAFARI_VERSION != TARGET_NOT_SUPPORTED || MIN_FIREFOX_VERSION != TARGET_NOT_SUPPORTED || ENVIRONMENT_MAY_BE_NODE
+#if !caniuse('js.Atomics.waitAsync') || MIN_CHROME_VERSION < 91
```
I guess MIN_CHROME_VERSION is still explicit here because of the above bug?

RReverser replied:
Yup. I raised issue on browser compat data repo, but for now it has to be explicit extra check.

sbc100 commented on emcc.py:
```diff
@@ -896,16 +899,6 @@ def parse_s_args(args):
 
-        # Special handling of browser version targets. A version -1 means that the specific version
-        # is not supported at all. Replace those with INT32_MAX to make it possible to compare e.g.
-        # #if MIN_FIREFOX_VERSION < 68
```
Maybe keep this comment on the new version of this code up on line 461?

RReverser replied:
Good catch.

sbc100 commented on tools/browser_compat_data_update.py:
```diff
@@ -0,0 +1,113 @@
+import json
```
`#!/usr/bin/env python3`

+ copyright 

+ docstring explaining what this script does.

sbc100 commented on tools/browser_compat_data_update.py:
```diff
@@ -0,0 +1,113 @@
+        return version[0]
+
+
+interesting_browsers = ["chrome", "firefox", "safari", "nodejs"]
```
We tend use single quote string in our python scripts.

RReverser replied:
Yeah I had them but then file got ugly so I ran Black autoformatter and it changed them to double quotes.

Do we have preferred autoformatter & config for Python files?

sbc100 replied:
I've used `yapf` in the past and there is the `.style.yapf` in the top level.

RReverser replied:
Thanks. I'll mark as unresolved and use it to reformat a bit later.

sbc100 commented on tools/browser_compat_data_update.py:
```diff
@@ -0,0 +1,113 @@
+        )
+
+with open(Path(__file__).parent / "browser_compat_data.json", "w") as f:
+    json.dump(out, f, indent="\t")
```
We use 2-space indentation in python scripts in this repo

sbc100 replied:
How about avoiding tabs in the json and using 2 or 4 spaces instead?

RReverser replied:
As for JSON, if would increase size somewhat more and I figured it might be not worth it.

sbc100 replied:
If you limit the json to just the features we care about that should no longer matter.

RReverser replied:
I'm really not a fan of that as it requires to remember to carefully add features to a separate list to avoid errors, and because it limits usefulness in 3rd-party libs...

RReverser replied:
Given the existing size of the repo, 100KB for checkout really seems like not a lot, and once checked in, it changes very rarely for existing features so much lower maintenance overhead.

sbc100 replied:
Its not just about the overall size, but also the ability to maintain and review.  For example, imagine when we update this from time to time and all kind of things will change.  I'd rather keep to limits and precise.

Third party library authors can request more over time, and if that happen more than N times a year we can reconsider.

The other nice thing about enumerating the full list of features we need is that the update process would fail right away if any feature is missing/renamed.

RReverser replied:
> For example, imagine when we update this from time to time and all kind of things will change.

Ideally you shouldn't review the auto-generated file. I mean, we could as well generate it as part of `bootstrap`, the only reason it's checked in is to avoid extra network requests.

> The other nice thing about enumerating the full list of features we need is that the update process would fail right away if any feature is missing/renamed.

That should never happen. A lot of tools and websites - caniuse, Babel, etc - depend on this database.

sbc100 reviewed (commented):
Pretty awesome!

sbc100 commented on tools/browser_compat_data_update.py:
```diff
@@ -0,0 +1,113 @@
+            support, normalized_browser
+        )
+
+with open(Path(__file__).parent / "browser_compat_data.json", "w") as f:
```
Maybe use `utils.write_file` here instead of the with block?

RReverser replied:
That would mean stringifying JSON in-memory first instead of writing to file in streaming way? Probably not a big difference, just wanted to point out it's not exactly the same.

sbc100 replied:
Limiting the size of the json should make that unimportant.

RReverser replied:
One problem I'm having when trying to import utils is:

```shell
> py .\tools\browser_compat_data_update.py
Traceback (most recent call last):
  File "C:\Users\me\Documents\emscripten\tools\browser_compat_data_update.py", line 5, in <module>
    from utils import write_file
  File "C:\Users\me\Documents\emscripten\tools\utils.py", line 12, in <module>
    from . import diagnostics
ImportError: attempted relative import with no known parent package
```

What am I doing wrong?

sbc100 replied:
See `tools/wasm-sourcemap.py` for an example:

```
__scriptdir__ = os.path.dirname(os.path.abspath(__file__))                                              
__rootdir__ = os.path.dirname(__scriptdir__)                                                                                                                                
sys.path.insert(0, __rootdir__)  
```

Then `from tools import utils`

sbc100 replied:
Also, perhaps this new tool should live in `tools/maint`?

RReverser replied:
> See `tools/wasm-sourcemap.py` for an example:

Hm is all of that really worth it just to avoid the `with` block?

> Also, perhaps this new tool should live in tools/maint?

I'm not very familiar with the folder structure here, but sure, I can move it.

sbc100 replied:
> > See `tools/wasm-sourcemap.py` for an example:
> 
> Hm is all of that really worth it just to avoid the `with` block?

probably not just for that.. but its nice if all our utility scripts have access to the same set of utility functions, and share the same standards.

> 
> > Also, perhaps this new tool should live in tools/maint?
> 
> I'm not very familiar with the folder structure here, but sure, I can move it.

I just noticed there are some other scripts that can/should be moved there too.

RReverser replied:
> probably not just for that.. but its nice if all our utility scripts have access to the same set of utility functions, and share the same standards.

But some other scripts like create_entry_points already don't do that, and also opt for `with` blocks. I thought that was intentional choice to avoid similar "glue".

sbc100 replied:
> > probably not just for that.. but its nice if all our utility scripts have access to the same set of utility functions, and share the same standards.
> 
> But some other scripts like create_entry_points already don't do that, and also opt for `with` blocks. I thought that was intentional choice to avoid similar "glue".

I don't feel strongly but over time it would be good to make all this consistent.

For what is worth I just add the sys.path boilerplate to create_entry_points as part of #20586

RReverser replied:
I'm not that deeply familiar with Python, but is there a way to declare all those scripts as a single package or something so it automatically allows such relative imports without duplicating the code? (figured I'd ask since the error talks about "parent package")

sbc100 replied:
Its complicated because when you run a script like that its running as `__main__` and not considered part of the `tools` package at that point.   It sadly doesn't just work as you might hope.  

You need to add the emscripten root to the sys.path as if the script was run from the top level (like emcc.py is).

Unless you want to go down the rabbit hole I suggest you just copy those three lines which update `sys.path` accordingly.

sbc100 replied:
Basically every script that wants to be directly runnable needs to do that little dance.

Ideally the PYTHON_PATH would already be setup to point the top level directly by the launching process..  but in this case we kind of want these files themselves to be directly runnable without creating extra launcher scripts.

RReverser replied:
That's a shame. Thanks for the explanation.

RReverser commented:
Somewhat related: I noticed a lot of checks for versions that seem ancient by now, e.g. in library_html5.js. I didn't bother updating those since I think those alternative codepaths should be simply cleaned up, but I don't know if Emscripten has official policy for "absolutely minimum browser versions we endeavour to support". Thoughts? @sbc100

kripken commented on src/library.js:
```diff
@@ -402,7 +402,7 @@ addToLibrary({
 #if (SHRINK_LEVEL < 2 || LINKABLE || process.env.EMCC_FORCE_STDLIBS) && !STANDALONE_WASM && !BULK_MEMORY
 
-#if MIN_CHROME_VERSION < 45 || MIN_EDGE_VERSION < 14 || MIN_FIREFOX_VERSION < 34 || MIN_IE_VERSION != TARGET_NOT_SUPPORTED || MIN_SAFARI_VERSION < 100101
+#if caniuse('js.TypedArray.copyWithin')
```
Shouldn't this be flipped? The old checks were "if we must support chrome before 45 etc. then add polyfill" but now it reads as "if I can use the feature then add polyfill".

RReverser replied:
Yup, good catch. I made similar mistake in other places before.

```suggestion
#if !caniuse('js.TypedArray.copyWithin')
```

kripken replied:
Sounds good. I can do a careful read to look for similar mistakes - is now a good time, or should I wait until it is no longer a draft?

RReverser replied:
I guess let me get around to addressing existing issues first, and then I'll re-request review. Since this changes a lot of files, extra eyes will certainly be welcome.

kripken commented on src/library.js:
```diff
@@ -402,7 +402,7 @@ addToLibrary({
 #if (SHRINK_LEVEL < 2 || LINKABLE || process.env.EMCC_FORCE_STDLIBS) && !STANDALONE_WASM && !BULK_MEMORY
 
-#if MIN_CHROME_VERSION < 45 || MIN_EDGE_VERSION < 14 || MIN_FIREFOX_VERSION < 34 || MIN_IE_VERSION != TARGET_NOT_SUPPORTED || MIN_SAFARI_VERSION < 100101
+#if caniuse('js.TypedArray.copyWithin')
```
Should all the comments below about version testing be removed, if we are going to trust the caniuse data source?

RReverser replied:
Yeah, I removed such comments in other places but probably missed here.

RReverser commented:
> Somewhat related: I noticed a lot of checks for versions that seem ancient by now, e.g. in library_html5.js. I didn't bother updating those since I think those alternative codepaths should be simply cleaned up, but I don't know if Emscripten has official policy for "absolutely minimum browser versions we endeavour to support". Thoughts?

@kripken do you have opinions on this? You mentioned in another PR that we want to support folks to run things "everywhere", but I think there still has to be some absolutely lowest limit to avoid maintaining hacks for versions of browsers that are no longer relevant forever.

I wonder where that limit lies, it would be good to document this somewhere.

sbc100 commented:
I wonder if there is some way we can avoid parsing the json when only compiling and not linking?

Could we have the `feature_matrix.py` include some kind of `init()` method which is called at the same point in `emcc.py` where we call `settings.limit_settings(None)`?

RReverser commented:
Hm but it auto-enables features like BULK_MEMORY too, aren't those needed during compilation?

sbc100 commented:
> Hm but it auto-enables features like BULK_MEMORY too, aren't those needed during compilation?

True... but I don't think MIN_XX_VERSION is valid during compilation so I don't think it would be useful in that case.

sbc100 commented:
> > Hm but it auto-enables features like BULK_MEMORY too, aren't those needed during compilation?
> 
> True... but I don't think MIN_XX_VERSION is valid during compilation so I don't think it would be useful in that case.

Yes, BULK_MEMORY is only auto-enabled during `phase_linker_setup` (my god that function is huge.. 1200 lines!)

RReverser commented:
Ah, so implicit bulk memory opt-in only changes implementation of functions like `memcpy` but not user code? I guess that's reasonable.

I'll take a look at addressing remaining issues somewhat later. As mentioned above, this was a bit of a fun weekend side project so it's a little bit lower down my pipeline.

sbc100 commented:
> Ah, so implicit bulk memory opt-in only changes implementation of functions like `memcpy` but not user code? I guess that's reasonable.

MIN_XX settings don't have any meaning when compiling:

```
$ emcc -c ~/test/hello.c -sMIN_CHROME_VERSION=60
emcc: warning: linker setting ignored during compilation: 'MIN_CHROME_VERSION' [-Wunused-command-line-argument]
```

If one wanted to opt into bulk memory operations at compile time one would currently need to add `-mbulk-memory`.

RReverser commented:
Right, yeah, that's what I meant by my question about implicit opt-in.

Although I feel like this begs another follow-up question... maybe they _should_ have effect at compile-time too? There's a couple of such Wasm features that apply at compile-time and that we could enable automatically under the right `MIN_XXX_VERSION` conditions.

sbc100 commented:
> Right, yeah, that's what I meant by my question about implicit opt-in.
> 
> Although I feel like this begs another follow-up question... maybe they _should_ have effect at compile-time too? There's a couple of such Wasm features that apply at compile-time and that we could enable automatically under the right `MIN_XXX_VERSION` conditions.

Maybe yes.. but I also really like the idea of keeping as many settings as link-only as possible.. so that the compile phase of emcc is basically just clang (as or near we can get). 

Within google we already use bare clang for the compiling and only use emcc for linking.. I'd love to move more in that direction if we can.

RReverser commented:
> Maybe yes.. but I also really like the idea of keeping as many settings as link-only as possible.. so that the compile phase of emcc is basically just clang (as or near we can get).

Heh, I remember a while back I was thinking the opposite actually: for example, a lot of optimisations in wasm-opt semantically don't need a finished linked executable, so if it could someday support relocatable Wasm objects, large projects could compile faster by running at least those optimisations on object files in parallel during the compilation phase rather than waiting for the entire linked thing. Aside from parallelisation, it would also improve speed by caching, as you don't need to re-run same optimisations on the already optimised object files.

That's unlikely to happen anytime soon though, just wanted to bring it up as a potential reason for doing more things compile-time instead.

sbc100 commented:
> > Maybe yes.. but I also really like the idea of keeping as many settings as link-only as possible.. so that the compile phase of emcc is basically just clang (as or near we can get).
> 
> Heh, I remember a while back I was thinking the opposite actually: for example, a lot of optimisations in wasm-opt semantically don't need a finished linked executable, so if it could someday support relocatable Wasm objects, large projects could compile faster by running at least those optimisations on object files in parallel during the compilation phase rather than waiting for the entire linked thing. Aside from parallelisation, it would also improve speed by caching, as you don't need to re-run same optimisations on the already optimised object files.
> 
> That's unlikely to happen anytime soon though, just wanted to bring it up as a potential reason for doing more things compile-time instead.

I think we have talked about that before too.   Thinking about it though its hard to see when it would be useful.. since fast/debug builds don't want to do wasm-opt at all and slow/release builds will likely want to do full-program optimization and don't mind paying the speed cost.   The method you describe here is kind of half way between the two... haveing said that, that is what thinLTO is and that is pretty popular.

RReverser commented:
> slow/release builds will likely want to do full-program optimization and don't mind paying the speed cost. The method you describe here is kind of half way between the two... haveing said that, that is what thinLTO is and that is pretty popular.

Yeah, exactly. Besides it's not really either-or: you can run all the optimisations that work on per-function/per-object basis in parallel, and then only those optimisations that need the whole program during link phase. This would be pretty similar to regular compilation + LTO pipeline where having link-time optimisations doesn't mean that objects themselves are not optimised at all.

sbc100 commented:
> > slow/release builds will likely want to do full-program optimization and don't mind paying the speed cost. The method you describe here is kind of half way between the two... haveing said that, that is what thinLTO is and that is pretty popular.
> 
> Yeah, exactly. Besides it's not really either-or: you can run all the optimisations that work on per-function/per-object basis in parallel, and then only those optimisations that need the whole program during link phase. This would be pretty similar to regular compilation + LTO pipeline where having link-time optimisations doesn't mean that objects themselves are not optimised at all.

Note that wasm-opt is already highly parallel internally so (unlike LLVM) you already a get a log of parallelism in the whole-program optimizer.   Doing it on a per-object-file basis could still be a win, but just FYI, the win is likely not as much as separate compilation is.

RReverser commented:
> Note that wasm-opt is already highly parallel internally so (unlike LLVM) you already a get a log of parallelism in the whole-program optimizer. Doing it on a per-object-file basis could still be a win, but just FYI, the win is likely not as much as separate compilation is.

Fair enough.
